### ISSUE TITLE
database locked when non commited transaction

### ISSUE THREAD

**Issue by PricelessRabbit** [mattn/go-sqlite3#856]
https://gist.github.com/PricelessRabbit/1e8a515364db5da19ee0d107b25928ce

If a transaction remains "pending" (no commit and no rollback), the db remains locked until the go process is restarted.
In the gist i created 2 different transaction. One of them (wrongTransaction) is not committed and there is no rollback. When de goroutine exits the reference to the transaction is lost, and it is impossilble to unlock the database (Process must be restarted). Are there any workarounds / solutions to ensure that db remains usable. 
Because in long running processes like a web server in case of some malfunctions of a transaction / deferred db close reboot the process is not an option.
I have the issue also in WAL / shared cache.
Maybe i'm missing something? it's possible to set a timer or something like that to delegate the sql driver to manage "lost" transactions?

**Comment by rittneje** [comment]
The best thing to do would be to `defer t.Rollback()` right after `db.Begin()`. That way it is impossible for the transaction to leak. `Rollback` is effectively a no-op if the transaction has already been rolled back or committed.

```go
tx, err := db.Begin()
if err != nil {
    return err
}
defer tx.Rollback()

...

return tx.Commit()
```

**Comment by PricelessRabbit** [comment]
> The best thing to do would be to `defer t.Rollback()` right after `db.Begin()`. That way it is impossible for the transaction to leak. `Rollback` is effectively a no-op if the transaction has already been rolled back or committed.
> 
> ```go
> tx, err := db.Begin()
> if err != nil {
>     return err
> }
> defer tx.Rollback()
> 
> ...
> 
> return tx.Commit()
> ```

@rittneje yes i implemented it in that way. But i'm worried if can be cases in which the defer rollback fails (i dont know if it is possible) and then there is no way to "recover" from the locked database.

**Comment by rittneje** [comment]
As far as I know, it is impossible for the rollback to ever truly fail. (That is, your database won't be left with some orphaned transaction in progress.) I would recommend posting on the [SQLite forum](https://sqlite.org/forum/forummain) if you want further clarification.

**Comment by mattn** [comment]
Closing as answered/stale: add 'defer tx.Rollback()' after Begin() (rollback after commit is a no-op); this is expected database/sql connection-lifecycle behavior rather than a driver bug, as explained in the thread.
